src/platform/mod.rs

```diff
@@ -48,19 +48,31 @@ macro_rules! impl_platform_host {
         ///
         /// This type may be constructed via the **host_from_id** function. **HostId**s may
         /// be acquired via the **ALL_HOSTS** const and the **available_hosts** function.
-        pub struct Host(HostInner);
+        // `Host` and `Device` cannot assume to be `Sync` as we have not yet been able to confirm
+        // whether or not the ASIO API is thread-safe.
+        //
+        // TODO: Try to contact ASIO to get more information. Review the existing implementation of
+        // the `asio` backend's `Host` and `Driver` types and see if the existing `Arc`s and
+        // `Mutex`s don't already make the API `Sync`.
+        pub struct Host(HostInner, crate::platform::NotSyncAcrossAllPlatforms);
```

Comment:
Hmm, so I think this should be one of the following instead:
- `!Send` and `!Sync` because we use Arc and the handles end up being shared between `Host` and `Device`s
- Just assume it's thread safe because we already wrap almost everything inside a Mutex.

Reply:
After a closer review it seems that the higher-level API exposed via the asio crate does indeed force synchronised access to the API and state transitions via the `Mutex` - I noticed some parts that weren't covered properly and patched them up.